src/data/sqlite_converter.py

```diff
@@ -0,0 +1,389 @@
+                gcd_root = os.path.join(root,file)
+                gcds_root.append(os.path.join(root,file))
+    if gcd_root == None:
+        gcd_root = gcd_rescue
+    for k in range(len(i3files_root) - len(gcds_root)):
+        gcds_root.append(gcd_root)
+    files_list.extend(i3files_root)
+    GCD_list.extend(gcds_root)
+    for folder in folders:
+        sub_root, sub_folders, sub_folder_files = next(os.walk(os.path.join(root,folder)))
+        gcds_folder = []
+        gcd_folder = None
+        i3files_folder = []
+        for sub_folder_file in sub_folder_files:
+            if has_extension(sub_folder_file, extensions):
+                if isi3(sub_folder_file):
+                    i3files_folder.append(os.path.join(sub_root,sub_folder_file))
+                else:
+                    gcd_folder = os.path.join(sub_root,sub_folder_file)
+                    gcds_folder.append(os.path.join(sub_root,sub_folder_file))
+        if gcd_folder == None:
+            gcd_folder = gcd_rescue
+        for k in range(len(i3files_folder) - len(gcds_folder)):
+            gcds_folder.append(gcd_folder)
+        files_list.extend(i3files_folder)
+        GCD_list.extend(gcds_folder)
+    
+    files_list, GCD_list = pairwiseshuffle(files_list, GCD_list)
+    return files_list, GCD_list
+
+def pairwiseshuffle(files_list, gcd_list):
+    df = pd.DataFrame({'i3': files_list, 'gcd': gcd_list})
+    df_shuffled = df.sample(frac = 1)
+    i3_shuffled = df_shuffled['i3'].tolist()
+    gcd_shuffled = df_shuffled['gcd'].tolist()
+    return i3_shuffled, gcd_shuffled
+
+
+def find_files(paths,outdir,db_name,gcd_rescue, extensions = None):
+    print('Counting files in: \n%s\n This might take a few minutes...'%paths)
+    if extensions == None:
+        extensions = ("i3.bz2",".zst",".gz")
+    input_files_mid = []
+    input_files = []
+    files = []
+    gcd_files_mid = []
+    gcd_files = []
+    for path in paths:
+        input_files_mid, gcd_files_mid = walk_directory(path, extensions, gcd_rescue)
+        input_files.extend(input_files_mid)
+        gcd_files.extend(gcd_files_mid)
+
+    if len(input_files) > 0:
+        input_files, gcd_files = pairwiseshuffle(input_files, gcd_files)
+        save_filenames(input_files, outdir, db_name)
+    return input_files, gcd_files
+    
+def save_filenames(input_files,outdir, db_name):
+    create_out_directory(outdir + '/%s/config'%db_name)
+    input_files = pd.DataFrame(input_files)
+    input_files.columns = ['filename']
+    input_files.to_csv(outdir + '/%s/config/i3files.csv'%db_name)
+    return
+
+def create_out_directory(outdir):
+    try:
+        os.makedirs(outdir)
+        return False
+    except:
+        print(' !!WARNING!! \n \
+            %s already exists. \n \
+            ABORTING! '%outdir)
+        return True
+
+def isempty(features):
+    if features['dom_x'] != None:
+        return False
+    else:
+        return True
+
+def extract_column_names(tmp_path, db_files, pulsemap):
```

Comment:
From here and down to L323 seems to be functionality that is specific to the `SQLiteDataConverter` class; if so I would recommend moving these to (private) class methods.

Reply:
Two distinct functionalities are in those lines: 

1) The creation of temporary .db-files which are done in parallel
2) The merging of said temporary databases files into a single, final database.

I cannot for the love of god put a private function (e.g. a function defined within the class) as an argument for map_asyinc

```
if __name__ == 'sqlite_converter':
              #parallel_extraction(settings[0])
              print('starting pool!')
              p = Pool(processes = workers)
              p.map_async(parallel_extraction, settings)
              p.close()
              p.join()
              print('pool')
              self._merge_databases()
```

When I tried to make `parallel_extraction` a private function, the script class would act as if it simply never called the function. However, this isn't whats going on; the function call for every worker fails (without error messages thx to multiprocessing). From googling I found that this is likely due to `self` not being handled correctly. If we were to obtain the error message, I'd think it would be something like:  `parallel_extraction recieved two arguments but expected 1`.

I **cannot** fix this. If we want it fixed someone needs to come up with a good idea.

Rasmus

Reply:
No worries, and thanks for the detailed description. I have created #11 and will try to work on this problem there.